driver/attributes.h

```diff
@@ -45,80 +44,85 @@ inline bool AttributeContainer::hasAttrAs<std::string>(int attr) const {
+            if constexpr (std::is_same_v<T, std::string>)
+                return value;
+            else {
+                const auto num_value = fromString<std::intptr_t>(value);
+                if constexpr (std::is_pointer_v<T>)
+                    return reinterpret_cast<T>(num_value);
+                else
+                    return static_cast<T>(num_value);
+            }
+        }
+        else {
+            if constexpr (std::is_same_v<T, std::string>)
+                return std::to_string(value);
+            else if constexpr (std::is_pointer_v<T>)
+                return reinterpret_cast<T>(value);
+            else
+                return static_cast<T>(value);
+        }
+    }, it->second);
 }
 
-template <>
-inline std::string AttributeContainer::getAttrAs<std::string>(int attr, const std::string & def) const {
-    auto it = strings.find(attr);
-    return (it == strings.end() ?
-        def : it->second);
+template <typename T>
+inline void AttributeContainer::setAttrSilent(int attr, const T& value) {
+    if constexpr (std::is_pointer_v<T>)
+        attributes.insert_or_assign(attr, reinterpret_cast<std::intptr_t>(value));
+    else
+        attributes.insert_or_assign(attr, value);
 }
 
 template <typename T>
-inline T AttributeContainer::setAttrSilent(int attr, const T& value) {
-    std::int64_t old_value = 0;
-    strings.erase(attr);
-    auto it = integers.find(attr);
-    if (it == integers.end()) {
-        integers.emplace(attr, (std::int64_t)value);
-    }
-    else {
-        old_value = it->second;
-        it->second = (std::int64_t)value;
-    }
-    return (T)old_value;
-}
+inline void AttributeContainer::setAttr(int attr, const T& value) {
+    const auto it = attributes.find(attr);
 
-template <>
-inline std::string AttributeContainer::setAttrSilent<std::string>(int attr, const std::string& value) {
-    std::string old_value;
-    integers.erase(attr);
-    auto it = strings.find(attr);
-    if (it == strings.end()) {
-        strings.emplace(attr, value);
-    }
-    else {
-        old_value = std::move(it->second);
-        it->second = value;
-    }
-    return old_value;
-}
+    if (it == attributes.end()) {
+        if constexpr (std::is_pointer_v<T>)
+            attributes.emplace(attr, reinterpret_cast<std::intptr_t>(value));
+        else
+            attributes.emplace(attr, value);
 
-template <typename T>
-inline T AttributeContainer::setAttr(int attr, const T& value) {
-    std::int64_t old_value = 0;
-    strings.erase(attr);
-    auto it = integers.find(attr);
-    if (it == integers.end()) {
-        integers.emplace(attr, (std::int64_t)value);
         onAttrChange(attr);
     }
     else {
-        old_value = it->second;
-        it->second = (std::int64_t)value;
-        if (old_value != (std::int64_t)value)
-            onAttrChange(attr);
-    }
-    return (T)old_value;
-}
+        const bool changed = std::visit(
```

Comment:
Would it make sense to rename that from `changed` to `is_different_value` ?

Reply:
Frankly, I don't see how it helps better understanding.

Reply:
Esp. changed **type** is also reflected there.

Reply:
agree